### PR TITLE
Add support for more operations

### PR DESCRIPTION
**Pull request by Robertorosmaninho**
This PR introduce support for: 

- IfThenElse
- For
- While
- Assign
- BitOr, BitAnd
- And, Or
- LTE, GTE, LT, GT, EQ, NEQ
- Indices property
- Min, Max
- Cast
- Case
- Comment, BlankLine

This PR also introduces 2 tests:  sum with sparse tensors and sum with dense tensors

### CODE REVIEW

File: dense_sum_test.cc [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include "include/taco.h"
+#include "include/taco/llvm.h"
+
```
Comment (guilhermeleobas) [inline dense_sum_test.cc]: Can you move this file to `llvm-include/` folder?

File: sparse_sum_test.cc [hunk 2]
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include "include/taco.h"
+#include "include/taco/llvm.h"
+
```
Comment (guilhermeleobas) [inline sparse_sum_test.cc]: Same! Move this file to `llvm-include/` folder

File: src/codegen/codegen_llvm.cpp [hunk 3]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
 
 void CodeGen_LLVM::visit(const Min* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "Min");
-  throw logic_error("Not Implemented for Min.");
+
+  if (op->operands.size() == 1) {
+    value = codegen(op->operands[0]);
+    return;
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Does this case happen in any test case? If not, add a `taco_iassert (op->operands.size() == 2)`

File: src/codegen/codegen_llvm.cpp [hunk 4]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+        (op->type.isInt()) ? Builder->CreateICmpSLT(a, b) : Builder->CreateICmpULT(a, b);
+    value = Builder->CreateSelect(icmp, a, b);
+
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      auto c = codegen(op->operands[i]);
+      icmp =
+          (op->type.isInt()) ? Builder->CreateICmpSLT(value, c) : Builder->CreateICmpULT(value, c);
+      value = Builder->CreateSelect(icmp, value, c);
+    }
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Can `op->operands.size()` be greater than 2? If not, you can remove this code

File: src/codegen/codegen_llvm.cpp [hunk 5]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+        (op->type.isInt()) ? Builder->CreateICmpSGT(a, b) : Builder->CreateICmpUGT(a, b);
+    value = Builder->CreateSelect(icmp, a, b);
+
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      auto c = codegen(op->operands[i]);
+      icmp =
+          (op->type.isInt()) ? Builder->CreateICmpSGT(value, c) : Builder->CreateICmpUGT(value, c);
+      value = Builder->CreateSelect(icmp, value, c);
+    }
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Same here!

File: src/codegen/codegen_llvm.cpp [hunk 6]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+
+  if (op->type.isFloat()) {
+    if (auto* FPa = llvm::dyn_cast<llvm::ConstantFP>(a)) {
+      if (auto* FPb = llvm::dyn_cast<llvm::ConstantFP>(b)) {
+        if (FPa->isNaN() || FPb->isNaN())
+          value = Builder->CreateFCmpUEQ(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOEQ(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: If neither operands are `llvm::ConstantFP`, we don't generate any code. You need to add an else branch that dispatch a floating-point comparison when both operands are not `ConstantFP`.

File: src/codegen/codegen_llvm.cpp [hunk 7]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+    value = codegen(op->operands[0]);
+    return;
+  }
+
+  if (op->type.isFloat()) {
+    // LLVM's minnum intrinsic only does binary ops
+    value = Builder->CreateMinNum(codegen(op->operands[0]), codegen(op->operands[1]));
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      value = Builder->CreateMinNum(value, codegen(op->operands[i]));
+    }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp:386]: Is `op->operands.size()` greater than 2 in any test case? If not, you can remove this code.

File: src/codegen/codegen_llvm.cpp [hunk 8]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
-  throw logic_error("Not Implemented for Min.");
+
+  if (op->operands.size() == 1) {
+    value = codegen(op->operands[0]);
+    return;
+  }
+
+  if (op->type.isFloat()) {
+    // LLVM's minnum intrinsic only does binary ops
+    value = Builder->CreateMinNum(codegen(op->operands[0]), codegen(op->operands[1]));
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: If you generate both `a` and `b` before this if condition, this code can be rewritten to:
```C++
auto* a = codegen(op->operands[0]);
auto* b = codegen(op->operands[1]);
if (op->type.isFloat()) {
    value = Builder->CreateMinNum(a, b);
} else if (op->type.isInt()) {
    ...
}
```

File: src/codegen/codegen_llvm.cpp [hunk 9]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+      if (auto* FPb = llvm::dyn_cast<llvm::ConstantFP>(b)) {
+        if (FPa->isNaN() || FPb->isNaN())
+          value = Builder->CreateFCmpUEQ(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOEQ(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
+  } else {
+    value = Builder->CreateICmpEQ(a, b);
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Change the else statement to check if `op->type.isInt() || op->type.isUInt()` is true
```suggestion
  } else if (op->type.isInt() || op->type.isUInt()) {
    value = Builder->CreateICmpEQ(a, b);
  } else {
     // raise an error here
  }
```

File: src/codegen/codegen_llvm.cpp [hunk 10]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+
+  if (op->type.isFloat()) {
+    if (auto* FPa = llvm::dyn_cast<llvm::ConstantFP>(a)) {
+      if (auto* FPb = llvm::dyn_cast<llvm::ConstantFP>(b)) {
+        if (FPa->isNaN() || FPb->isNaN())
+          value = Builder->CreateFCmpONE(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpONE(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Same here! You don't generate any code if `a` and `b` are not `ConstantFP`.

File: src/codegen/codegen_llvm.cpp [hunk 11]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+  } else {
+    value = Builder->CreateICmpNE(a, b);
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Same here! Ensure `op` type is either `Int` or `UInt`

File: src/codegen/codegen_llvm.cpp [hunk 12]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+          value = Builder->CreateFCmpUGT(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOGT(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
+  } else if (op->type.isUInt()) {
+    value = Builder->CreateICmpUGT(a, b);
+  } else {
+    value = Builder->CreateICmpSGT(a, b);
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Same

File: src/codegen/codegen_llvm.cpp [hunk 13]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+          value = Builder->CreateFCmpULT(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOLT(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
+  } else if (op->type.isUInt()) {
+    value = Builder->CreateICmpULT(a, b);
+  } else {
+    value = Builder->CreateICmpSLT(a, b);
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Same

File: src/codegen/codegen_llvm.cpp [hunk 14]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+          value = Builder->CreateFCmpUGE(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOGE(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
+  } else if (op->type.isUInt()) {
+    value = Builder->CreateICmpUGE(a, b);
+  } else {
+    value = Builder->CreateICmpSGE(a, b);
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Same

File: src/codegen/codegen_llvm.cpp [hunk 15]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+          value = Builder->CreateFCmpULE(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOLE(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
+  } else if (op->type.isUInt()) {
+    value = Builder->CreateICmpULE(a, b);
+  } else {
+    value = Builder->CreateICmpSLE(a, b);
+  }
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Same

File: src/codegen/codegen_llvm.cpp [hunk 16]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+  Builder->SetInsertPoint(true_bb);
+  auto b = codegen(op->b);
+  Builder->CreateBr(end_bb);
+
+  // Create the PHI node on the end_bb to get the result from a && b
+  Builder->SetInsertPoint(end_bb);
+  auto phi = Builder->CreatePHI(get_int_type(1, *this->Context), 2);
+  phi->addIncoming(a, actual_bb);
+  phi->addIncoming(b, true_bb);
+  value = phi;
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Can you implement this instruction as a function call? For instance, create a function in the module that implements a `logical and`, let's call this function `logical_and`. Whenever you emit code for the `And*` opcode, you emit a call to the `logical_and` function.

File: src/codegen/codegen_llvm.cpp [hunk 17]
```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+  Builder->SetInsertPoint(false_bb);
+  auto b = codegen(op->b);
+  Builder->CreateBr(end_bb);
+
+  // Create the PHI node on the end_bb to get the result from a || b
+  Builder->SetInsertPoint(end_bb);
+  auto phi = Builder->CreatePHI(get_int_type(1, *this->Context), 2);
+  phi->addIncoming(a, actual_bb);
+  phi->addIncoming(b, false_bb);
+  value = phi;
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: The same for `logical_or`. Given that both functions are sort of similar, I think you can reuse almost the same code to implement both functions.

File: src/codegen/codegen_llvm.cpp [hunk 18]
```diff
@@ -390,12 +601,49 @@ void CodeGen_LLVM::visit(const Call* op) {
...
+    return !alwaysMatch ? clauses[0].second
+                        : IfThenElse::make(clauses[0].first, clauses[0].second, Comment::make(""));
+  } else {
+    return IfThenElse::make(
+        clauses[0].first, clauses[0].second, caseToIfThenElse(rest, alwaysMatch));
+  }
+}
+}  // anonymous namespace
+
+void CodeGen_LLVM::visit(const Case* op) {  // Not tested
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp:718]: Given we don't have any tests for this opcode, let's postpone it to another PR. Just comment the call to `codegen(caseToIfThenElse(...))` and throw an error.

File: src/codegen/codegen_llvm.cpp [hunk 19]
```diff
@@ -484,7 +732,31 @@ void CodeGen_LLVM::visit(const For* op) {
...
+  // entry condition
+  Builder->SetInsertPoint(header);
+  auto condition = codegen(op->cond);
+  Builder->CreateCondBr(condition, loop, exit);
+
+  // codegen body
+  Builder->SetInsertPoint(loop);
+  codegen(op->contents);
+
+  // create unconditional branch to check
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]:
```suggestion
  // create unconditional branch to header
```

File: src/codegen/codegen_llvm.cpp [hunk 20]
```diff
@@ -484,7 +732,31 @@ void CodeGen_LLVM::visit(const For* op) {
...
+  Builder->CreateBr(header);
+
+  // seet the insert point for the exit loop
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]:
```suggestion
  // set the insert point for the exit loop
```

File: src/codegen/codegen_llvm.cpp [hunk 21]
```diff
@@ -622,23 +904,37 @@ void CodeGen_LLVM::visit(const Yield* op) {
...
+    } else {
+      auto size = this->Builder->CreateMul(
+        sext_num_elements, get_int_constant(64, 4, *this->Context));
+
+     auto call = emitExternalCall("malloc", voidptr, {i64}, {size});
+     call->setName("calloc.ret");
+
+     auto bitcast = this->Builder->CreateBitCast(call, i32p);
+     this->Builder->CreateStore(bitcast, alloca);
+    }*/
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Let's do this in a later PR. Can you emit a warning here saying we're not emitting code for `Allocate`

File: src/codegen/codegen_llvm.cpp [hunk 22]
```diff
@@ -648,11 +944,11 @@ void CodeGen_LLVM::visit(const Free* op) {
 
 void CodeGen_LLVM::visit(const Comment* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "Comment");
-  throw logic_error("Not Implemented for Comment.");
+  // no op, do nothing
 }
 
 void CodeGen_LLVM::visit(const BlankLine* op) {
-  auto _ = CodeGen_LLVM::IndentHelper(this, "BlankLine");
+  // no-op, do nothing
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp:997]:
```suggestion
  auto _ = CodeGen_LLVM::IndentHelper(this, "BlankLine");
  // no-op, do nothing
```

File: src/codegen/codegen_llvm.cpp [hunk 23]
```diff
@@ -715,7 +1012,19 @@ void CodeGen_LLVM::visit(const GetProperty* op) {
...
+      auto* index = get_int_constant(32, op->index, *this->Context);
+
+      auto* load1 = this->Builder->CreateLoad(
+          this->Builder->CreateStructGEP(tensor, (int) TensorProperty::Indices));  // i8***
+      auto* load2 =
+          this->Builder->CreateLoad(this->Builder->CreateInBoundsGEP(load1, mode));  // i8**
+      auto* load3 =
+          this->Builder->CreateLoad(this->Builder->CreateInBoundsGEP(load2, index));  // i8*
+      value = this->Builder->CreateBitCast(load3, i32p);
+      break;
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: Let's move this to a function. This might make things easier to debug

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi @Robertorosmaninho, your PR looks good. I can confirm both tests are running fine, the segfault you see after the code is executed is not related to your PR, and it was introduced in #10. 

I've made an initial review. Let's move the implementation of `logical_and` and `logical_or` to a function and emit a function call to it. Also, I guess the implementation for the comparison opcodes can be merged into a single function.

Let me know if you have any questions.

File: src/codegen/codegen_llvm.cpp (hunk 4 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Yes! This is an example: [A(i) = B(i) + C(i) + D(i)](http://tensor-compiler.org/codegen.html?expr=A(i)%20=%20B(i)%20+%20C(i)%20+%20D(i)&format=A:s:0;B:s:0;C:s:0;D:s:0)

``` C
 while ((iB < pB1_end && iC < pC1_end) && iD < pD1_end) {
    int32_t iB0 = B1_crd[iB];
    int32_t iC0 = C1_crd[iC];
    int32_t iD0 = D1_crd[iD];
    int32_t i = TACO_MIN(iB0,TACO_MIN(iC0,iD0));
    if ((iB0 == i && iC0 == i) && iD0 == i) {
      A_vals[iA] = (B_vals[iB] + C_vals[iC]) + D_vals[iD];
      iA++;
    }
...
```

File: src/codegen/codegen_llvm.cpp (hunk 3 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: I couldn't generate this example, so I'll add a `taco_iassert (op->operands.size() >= 2)`

File: dense_sum_test.cc (hunk 1 above)
Comment (Robertorosmaninho) [reply dense_sum_test.cc]: `llvm-include/` or `llvm-examples/` ?

File: src/codegen/codegen_llvm.cpp (hunk 19 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Ok!

File: src/codegen/codegen_llvm.cpp (hunk 20 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: ok!

File: src/codegen/codegen_llvm.cpp (hunk 21 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Should I add a `throw logic_error("Not Implemented for Allocate.");` after `auto _ = CodeGen_LLVM::IndentHelper(this, "Warning: Missing Allocate implementation");` ?

File: src/codegen/codegen_llvm.cpp (hunk 22 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp:997]: ok!

File: dense_sum_test.cc (hunk 1 above)
Comment (guilhermeleobas) [reply dense_sum_test.cc]: Sorry, `llvm-examples/`

File: sparse_sum_test.cc (hunk 2 above)
Comment (guilhermeleobas) [reply sparse_sum_test.cc]: `llvm-examples/`**

File: src/codegen/codegen_llvm.cpp (hunk 4 above)
Comment (guilhermeleobas) [reply src/codegen/codegen_llvm.cpp]: Notice that `TACO_MIN` only take two arguments. I don't think there's a case where `min` takes a variable number of arguments.

File: src/codegen/codegen_llvm.cpp (hunk 21 above)
Comment (guilhermeleobas) [reply src/codegen/codegen_llvm.cpp]: No, use `taco_uwarning` for the warning message

```c++
 taco_uwarning << "Missing LLVM implementation for Allocate opcode" << std::endl;
```

Don't throw a logic_error

File: src/codegen/codegen_llvm.cpp (hunk 4 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Yes, but this [codegen](https://github.com/Quansight-Labs/taco/blob/ebceca71c8181031449da164b42e63ebc6be25e1/src/codegen/codegen_llvm.cpp#L350) doesn't call min again to generate `TACO_MIN(iC0,iD0)`, instead this is the first comparison and then it compares with `iB0`.

File: src/codegen/codegen_llvm.cpp (hunk 6 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Should I assume that either operand may be a QNAN then?

File: src/codegen/codegen_llvm.cpp (hunk 8 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: ok!

File: src/codegen/codegen_llvm.cpp (hunk 6 above)
Comment (guilhermeleobas) [reply src/codegen/codegen_llvm.cpp]: Yes! But you see the "problem" in the code right now? TACO only generates code iff one (or both) of the operands are `ConstantFP`

File: src/codegen/codegen_llvm.cpp (hunk 16 above)
Comment (guilhermeleobas) [reply src/codegen/codegen_llvm.cpp]: Ignore this comment. Doesn't make sense to create a function for this opcode.

File: src/codegen/codegen_llvm.cpp (hunk 17 above)
Comment (guilhermeleobas) [reply src/codegen/codegen_llvm.cpp]: Ignore this comment. Doesn't make sense to create a function for this opcode.

File: src/codegen/codegen_llvm.cpp [hunk 24]
```diff
@@ -320,67 +333,269 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+
+  // Create the BasicBlocks
+  auto* false_bb = llvm::BasicBlock::Create(*this->Context, "false_or_bb", this->Func);
+  auto* end_bb = llvm::BasicBlock::Create(*this->Context, "end_or_bb", this->Func);
+
+  // Generate first condition
+  auto a = codegen(op->a);
+  Builder->CreateCondBr(a, false_bb, end_bb);
+
+  // If false -> Generate the sencond condition to test
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]:
```suggestion
  // If false -> Generate the second condition to test
```

File: src/codegen/codegen_llvm.cpp (hunk 21 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: `taco_uwarning` stops the code generation. Are ther any other instruction that issue a warning without interrupting the code generation process?

File: src/codegen/codegen_llvm.cpp (hunk 23 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: ok!

File: src/codegen/codegen_llvm.cpp (hunk 24 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: ok!

File: src/codegen/codegen_llvm.cpp (hunk 6 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Solved!

File: src/codegen/codegen_llvm.cpp (hunk 10 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Done!

File: src/codegen/codegen_llvm.cpp (hunk 12 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Done!

File: src/codegen/codegen_llvm.cpp (hunk 14 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Done!

File: src/codegen/codegen_llvm.cpp (hunk 15 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Done!

File: src/codegen/codegen_llvm.cpp (hunk 18 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp:718]: Actually, this call is essential to nested IfThenElse!
This is not the `case` used for Switch, this is used to generate a series of conditionals like [here](https://github.com/Robertorosmaninho/taco/blob/4afa91a6667be480ee05de2f8ee1bb746d909b36/include/taco/ir/ir.h#L574)!

File: src/codegen/codegen_llvm.cpp (hunk 9 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Done!

File: src/codegen/codegen_llvm.cpp (hunk 11 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp]: Done!

File: src/codegen/codegen_llvm.cpp (hunk 21 above)
Comment (guilhermeleobas) [reply src/codegen/codegen_llvm.cpp]: Apply the following diff

```diff
diff --git a/src/error.cpp b/src/error.cpp
index 9be355db..90292a45 100644
--- a/src/error.cpp
+++ b/src/error.cpp
@@ -62,7 +62,8 @@ ErrorReport::ErrorReport(const char *file, const char *func, int line,
 void ErrorReport::explodeWithException() {
   TacoException e = TacoException(msg->str());
   delete msg;
-  throw e;
+  if (!warning)
+    throw e;
 }

 }
```

File: src/codegen/codegen_llvm.h [hunk 25]
```diff
@@ -45,13 +45,15 @@ class CodeGen_LLVM : public CodeGen {
   void pushSymbol(const std::string& name, llvm::Value* v);
   void removeSymbol(const std::string& name);
   llvm::Value* getSymbol(const std::string& name);
+  bool containsSymbol(const std::string& name);
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.h:51]: Add the following
```suggestion
  bool containsSymbol(const std::string& name);
  llvm::Function* getOrInsertGetIndicesFunction();
```

File: src/codegen/codegen_llvm.cpp [hunk 26]
```diff
@@ -320,67 +372,299 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+        else
+          value = Builder->CreateFCmpOEQ(a, b);  // OEQ means that neither operand is a QNAN
+      } else {
+        value = Builder->CreateFCmpUEQ(a, b);
+      }
+    } else {
+      value = Builder->CreateFCmpUEQ(a, b);
+    }
+  } else {
+    taco_iassert(op->type.isInt() || op->type.isUInt()) << "op->type is " << op->type;
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]: `op->type` can be boolean here as well
```suggestion
    taco_iassert(op->type.isInt() || op->type.isUInt() || op->type.isBool()) << "op->type is " << op->type;
```

File: src/codegen/codegen_llvm.cpp [hunk 27]
```diff
@@ -320,67 +372,299 @@ void CodeGen_LLVM::visit(const Rem* op) {
...
+        else
+          value = Builder->CreateFCmpONE(a, b);  // ONE means that neither operand is a QNAN
+      } else {
+        value = Builder->CreateFCmpUNE(a, b);
+      }
+    } else {
+      value = Builder->CreateFCmpUNE(a, b);
+    }
+  } else {
+    taco_iassert(op->type.isInt() || op->type.isUInt()) << "op->type is " << op->type;
```
Comment (guilhermeleobas) [inline src/codegen/codegen_llvm.cpp]:
```suggestion
    taco_iassert(op->type.isInt() || op->type.isUInt() || op->type.isBool()) << "op->type is " << op->type;
```

File: src/codegen/codegen_llvm.cpp (hunk 18 above)
Comment (guilhermeleobas) [reply src/codegen/codegen_llvm.cpp:718]: Ok! But do we have any test the exercise the code we generate?

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi Roberto!

The PR looks good. After these fixes, we can merge it. Thanks for all the work you've been doing.

**Comment by Robertorosmaninho** [comment]
https://github.com/Quansight-Labs/taco/pull/12#discussion_r671435995 Well, this function doesn't create any new code just call IfThenElse for IF-ELSEIF-ELSE cases. In this example [A(i) = B(i) + C(i)](http://tensor-compiler.org/codegen.html?expr=A(i)%20=%20B(i)%20+%20C(i)&format=A:s:0;B:s:0;C:s:0) this function is called one time for `assemble` and another for `compute` to generate code for 
```
if (iB0 == i && iC0 == i) {
      A_vals[iA] = B_vals[iB] + C_vals[iC];
      iA++;
    }
    else if (iB0 == i) {
      A_vals[iA] = B_vals[iB];
      iA++;
    }
    else {
      A_vals[iA] = C_vals[iC];
      iA++;
    }
```
If the sum of the tensor has more operands, then more `else if` are needed and this function is called many times!
LLVM IR for A(i) = B(i) + C(i) with debug output (look for "Case"): https://pastebin.com/TaDmbjDj - Example: https://pastebin.com/b56gy5NV
LLVM IR for A(i) = B(i) + C(i) + D(i) with debug output (look for "Case"): [gist](https://gist.github.com/Robertorosmaninho/b9a20a56225bbfa0bf839ef6289bec08) - Example: https://pastebin.com/tND6Q8iw

File: src/codegen/codegen_llvm.cpp (hunk 7 above)
Comment (Robertorosmaninho) [reply src/codegen/codegen_llvm.cpp:386]: Yes!

File: src/error.cpp [hunk 28]
```diff
@@ -13,12 +13,21 @@ const char* TacoException::what() const noexcept {
...
+                         Kind kind,
+                         bool warning)
+    : msg(NULL)
+    , file(file)
+    , func(func)
+    , line(line)
+    , condition(condition)
+    , conditionString(conditionString)
+    , kind(kind)
+    , warning(warning) {
```
Comment (guilhermeleobas) [inline src/error.cpp]: Can you undo any change related to clang-format in this file?

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi @Robertorosmaninho, your PR looks almost ready to be merged. Just a small request and we can merge it.

File: src/error.cpp (hunk 28 above)
Comment (Robertorosmaninho) [reply src/error.cpp]: Yes!

**Comment by guilhermeleobas** [comment]
Thanks, @Robertorosmaninho

**Pull request merged**